added idle action for climate

danielkucera opened the pull request:
## Description:


**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

```
[01:39:01][VV][api.service:210]: send_climate_state_response: ClimateStateResponse {
  key: 3913343285
  mode: CLIMATE_MODE_HEAT
  current_temperature: 24
  target_temperature: 23.5
  target_temperature_low: 0
  target_temperature_high: 0
  away: NO
  action: CLIMATE_ACTION_IDLE
}
```

Commit:
added idle

OttoWinter commented on esphome/components/climate/climate_mode.h:
```diff
@@ -25,6 +25,8 @@ enum ClimateAction : uint8_t {
   CLIMATE_ACTION_COOLING = 2,
   /// The climate device is actively heating (usually in heat or auto mode)
   CLIMATE_ACTION_HEATING = 3,
+  /// The climate device is idle (applicable for any mode)
```
We should make more clear the difference between IDLE and OFF. A second sentence in this documentation comment would help.

OttoWinter reviewed (commented):
Update looks mostly good - just some more documentation would be good to avoid developers getting confused about the different actions.

This will also need an update in two places:
 - aioesphomeapi: https://github.com/esphome/aioesphomeapi
 - the Home Assistant ESPHome integration (add `idle` to the `map_enum`)

Commit:
more clear state description

danielkucera commented:
@OttoWinter is it more clear now? I can prepare patches for the api and hass afterwards.

OttoWinter commented:
It's good now 👍

Also, I just realized you also need to update a couple more places in the esphome codebase as well:

 - `ClimateCall::set_mode` - example https://github.com/esphome/esphome/pull/845/files#diff-61e21d4296f352a45513a2d419419e5cR116-R117
 - `ClimateTraits` - add a supports_idle option and update supports_mode(),  https://github.com/esphome/esphome/pull/845/files#diff-541e2f2d8267d7329c76aa961fcb4665R17, https://github.com/esphome/esphome/pull/845/files#diff-17271c6878bd9fbfe237dd78c248256eR24
 - update mqtt climate (optional, but low hanging fruit) - https://github.com/esphome/esphome/pull/845/files#diff-6b9079b16765b3d6fff78888d6a5505dR35

DAVe3283 commented:
This will be very welcome when it is working, especially now that Home Assistant is showing heating/cooling active state on the temperature graph. Currently, a heat- or cool-only climate component in ESPHome reports that it is continually heating or cooling even though it is not.

OttoWinter commented:
@DAVe3283 Please explain what you mean - this change won't affect anything in how HA displays things.

danielkucera commented:
@OttoWinter it might. This is how it displays idle heating: ![Screenshot_20191123-195934_Firefox](https://user-images.githubusercontent.com/1734256/69483793-085ec580-0e2c-11ea-9c2e-fc1b99d49b64.png)

DAVe3283 commented:
In recent HA builds, they have started showing the time a climate object is heating or cooling. This is the green vertical bars in this screenshot:
![hass1](https://user-images.githubusercontent.com/6894502/69483768-a5980a80-0de8-11ea-9e6d-c404953ef568.PNG)

For a ESPHome climate component, it shows "cooling" all the time (purple):
![hass2](https://user-images.githubusercontent.com/6894502/69483807-1dfecb80-0de9-11ea-8bba-cd2de6643b32.PNG)
Even though the chiller only ran for about an hour (the temperature dip).

OttoWinter committed:
Merge branch 'dev' into climate-idle

OttoWinter committed:
Also add drying/fan

Putting it in this PR because it will be put in the same aioesphomeapi release anyway.

OttoWinter committed:
Update bang_bang for idle action

danielkucera commented:
Oh, perfect, thanks for finishing this @OttoWinter.
